src-java/kilda-model/src/main/java/org/openkilda/model/BfdSession.java

```diff
@@ -15,83 +15,138 @@
+                      String remoteIpAddress, @NonNull Integer port, Integer discriminator) {
+        data = BfdSessionDataImpl.builder()
+                .switchId(switchId).ipAddress(ipAddress).remoteSwitchId(remoteSwitchId)
+                .remoteIpAddress(remoteIpAddress).port(port).discriminator(discriminator).build();
+    }
+
+    public BfdSession(@NonNull BfdSessionData data) {
+        this.data = data;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        BfdSession that = (BfdSession) o;
+        return new EqualsBuilder()
+                .append(getSwitchId(), that.getSwitchId())
+                .append(getIpAddress(), that.getIpAddress())
+                .append(getRemoteSwitchId(), that.getRemoteSwitchId())
+                .append(getRemoteIpAddress(), that.getRemoteIpAddress())
+                .append(getPort(), that.getPort())
+                .append(getDiscriminator(), that.getDiscriminator())
+                .isEquals();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getSwitchId(), getIpAddress(), getRemoteSwitchId(), getRemoteIpAddress(),
+                getPort(), getDiscriminator());
     }
 
     /**
-     * BfdSession builder with prefilled mandatory fields.
+     * Defines persistable data of the BfdSession.
      */
-    public static BfdSessionBuilder builder(@NonNull SwitchId switchId,
-                                            @NonNull Integer port) {
-        return new BfdSessionBuilder()
-                .switchId(switchId)
-                .port(port);
+    public interface BfdSessionData {
+        SwitchId getSwitchId();
+
+        void setSwitchId(SwitchId switchId);
+
+        String getIpAddress();
+
+        void setIpAddress(String ipAddress);
+
+        SwitchId getRemoteSwitchId();
+
+        void setRemoteSwitchId(SwitchId remoteSwitchId);
+
+        String getRemoteIpAddress();
+
+        void setRemoteIpAddress(String remoteIpAddress);
+
+        Integer getPort();
+
+        void setPort(Integer port);
+
+        Integer getDiscriminator();
+
+        void setDiscriminator(Integer discriminator);
+    }
+
+    /**
+     * POJO implementation of BfdSessionData.
+     */
+    @Data
+    @Builder
+    @NoArgsConstructor
+    @AllArgsConstructor
+    static final class BfdSessionDataImpl implements BfdSessionData, Serializable {
+        private static final long serialVersionUID = 1L;
+        @NonNull SwitchId switchId;
+        String ipAddress;
+        SwitchId remoteSwitchId;
+        String remoteIpAddress;
+        @NonNull Integer port;
+        Integer discriminator;
+    }
+
+    /**
+     * A cloner for BfdSession entity.
+     */
+    @Mapper
+    public interface BfdSessionCloner {
```

Comment:
Why you prefer interface as the base for mapper? If it will be abstract class, it will allow restricting access to method that can be restricted (there is not too many such methods now, but they exists). Also in opposite to weird syntax hack (default methods in interfaces), code in methods of abstract classes are natural.

Reply:
With interfaces it's a bit less verbose:
```
    public interface BfdSessionCloner {
        BfdSessionCloner INSTANCE = Mappers.getMapper(BfdSessionCloner.class);

        void copy(BfdSessionData source, @MappingTarget BfdSessionData target);

        default BfdSessionData copy(BfdSessionData source) {
            ...
        }
    }
```
vs
```
    public static abstract class BfdSessionCloner {
        public static final BfdSessionCloner INSTANCE = Mappers.getMapper(BfdSessionCloner.class);

        public abstract void copy(BfdSessionData source, @MappingTarget BfdSessionData target);

        public BfdSessionData copy(BfdSessionData source) {
            ...
        }
    }
```

No other preferences. Also, I found only 2 entities where we need non-public methods in a cloner - those can be changed to abstract classes.